### DISCUSSION TITLE
How to close a Snapshot after using it to read a file from disk cache ?

### DISCUSSION THREAD

**Discussion by LeMimit** [coil-kt/coil#1021] [Q&A] [answered]
Hello !

First, thanks for this amazing library.

I was trying to use the new `DiskCache` from the version `2.0.0-alpha05`. 

I saw in the documentation that calling `Snapshot.close` or `Snapshot.closeAndEdit` is really important.

```kotlin
    /**
     * Get the entry associated with [key].
     *
     * IMPORTANT: **You must** call either [Snapshot.close] or [Snapshot.closeAndEdit] when finished
     * reading the snapshot. An open snapshot prevents editing the entry or deleting it on disk.
     */
    operator fun get(key: String): Snapshot?
```

After, I was looking around and I saw that in the code of `HttpUriFetcher` which access to the `DiskCache` in some case neither `Snapshot.close` or `Snapshot.closeAndEdit` are called.

For example here 

```kotlin
                // Always return cached images with empty metadata as they were likely added manually.
                if (snapshot.metadata.length() == 0L) {
                    return SourceResult(
                        source = snapshot.toImageSource(),
                        mimeType = getMimeType(url, null),
                        dataSource = DataSource.DISK
                    )
                }
```

Is it a bug or did I not understand somethink ? Can you help me ?

Thanks !

**Comment by colinrtwhite** [comment, marked as answer]
Thanks! And yep, this is intentional as we need to keep the disk snapshot open until it's finished being read by a `Decoder`. Otherwise, the file in the disk cache could be deleted while it's being decoded. It's always eventually closed [here](https://github.com/coil-kt/coil/blob/d0cff5641158d5fbdc44c9ad151b9df08631461d/coil-base/src/main/java/coil/intercept/EngineInterceptor.kt#L139) even if there's an uncaught exception.

**Reply by LeMimit** [reply]
Thanks for the answer. Ok I understand now !
